### PR TITLE
feat: introduce `forwardRef` support

### PR DESCRIPTION
**Pull request by acao**
follow up for #49

upon writing tests, I noticed that my original version didn't support a pattern where someone calls `forwardRef` directly - so this takes care of multiple cases, but i think I can tweak this in a few ways to make it much better?

- setting the nodes names directly using a template string approach is less than ideal of course, but I couldn't get other approaches working, such as setting `typeParameters` directly. would be happy to improve this part
- the way I'm introspecting AST nodes for `forwardRef` - it may be more efficient to look for every `CallExpression` named `forwardRef`, i guess that would be `find()` and `filter()` on `NodePath`?
- the change to `getComponentName` I made could possibly have unintended side effects? 
  -  It allows all `ArrowFunctionExpressions` to potentially return their grandparent's node for a name if the arrow function is anonymous (as is the case with `forwardRef`)?
  - accidentally resolving invalid `componentName`s is fine, because they won't have prop types to match, so the codemod will ignore them.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,8 +52,11 @@ MyComponent.propTypes = {
   bar: PropTypes.string.isRequired,
   foo: PropTypes.number,
 }
+```
 
-// Output
+output:
```
Comment (mskelton) [inline README.md:57]:
```suggestion
Output:
```

File: README.md [hunk 2]
```diff
@@ -33,9 +35,11 @@ the left and instantly see the output on the right!
 
 [![Screenshot](web/screenshot.png?v=1)](https://mskelton.dev/ratchet)
 
-## Example Input/Output
+## Example: Function Component
 
-```ts
+input:
```
Comment (mskelton) [inline README.md:40]:
```suggestion
Input:
```

File: __testfixtures__/forward-ref-and-func.output.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import React, { forwardRef } from "react"
+
+interface MyComponentProps {
+  bar: string;
+  foo?: number;
+}
+
+export const MyComponent = forwardRef<React.ReactHTMLElement, MyComponentProps>((props, ref) => {
```
Comment (mskelton) [inline __testfixtures__/forward-ref-and-func.output.js:8]:
```suggestion
export const MyComponent = forwardRef<HTMLElement, MyComponentProps>((props, ref) => {
```

File: transform.ts [hunk 4]
```diff
@@ -241,6 +240,25 @@ function createInterface(path: NodePath, componentTypes: CollectedTypes) {
...
 }
+/**
+ * If forwardRef is being used, declare the props.
+ * Otherwise, return false
+ */
+function addForwardRefTypes(path: NodePath, typeName: string): boolean {
+  // for `React.forwardRef()`
+  let isForwardRef = path.node.callee?.property?.name === "forwardRef"
+  if (isForwardRef) {
+    path.node.callee.property.name = `forwardRef<React.ReactHTMLElement, ${typeName}>`
```
Comment (mskelton) [inline transform.ts:251]:
```suggestion
    path.node.callee.property.name = `forwardRef<HTMLElement, ${typeName}>`
```

File: transform.ts [hunk 5]
```diff
@@ -241,6 +240,25 @@ function createInterface(path: NodePath, componentTypes: CollectedTypes) {
...
+    return isForwardRef
+  }
+  // if calling `forwardRef()` directly
+  isForwardRef = path.node.callee?.name === "forwardRef"
```
Comment (mskelton) [inline transform.ts:255]: Let's cleanup this duplication by making a function that finds the forward ref node, then we can act on that node to add the types.

File: README.md [hunk 6]
```diff
@@ -33,9 +35,11 @@ the left and instantly see the output on the right!
...
+
+```jsx
```
[the remark this answers is not in the document]
Comment (acao) [reply README.md:42]: @mskelton ok! this just made things work better for me in vscode, but excellent point

File: transform.ts (hunk 5 above)
Comment (acao) [reply transform.ts:255]: this is better!

File: transform.ts (hunk 4 above)
Comment (acao) [reply transform.ts:251]: see my notes from the description

> setting the nodes names directly using a template string approach is less than ideal of course, but I couldn't get other approaches working, such as setting typeParameters directly. would be happy to improve this part

I tried setting `path.value.typeArguments` directly, but with identifiers only - I didn't try wrapping them in `j.genericTypeAnnotation(` which makes sense!

Comment (mskelton) [reply transform.ts:251]: > maybe I'm working from a different version of @types/react but I don't get an React.HTMLElement only React.ReactHTMLElement?

As I showed in my suggestion, the proper interface is `HTMLElement` without the `React.` prefix.  `HTMLElement` comes from TypeScript (lib.dom.ts) as it describes the actual `HTMLElement` interface as specified in the web spec.

- https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L6357
- https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement

**Comment by krishnaglick** [comment]
Heya, bumping this in case it's at all mergeable!

**Comment by mskelton** [comment]
@krishnaglick This is not mergeable yet as there are a number of unresolved comments and changes.

**Comment by krishnaglick** [comment]
> @krishnaglick This is not mergeable yet as there are a number of unresolved comments and changes.

I wasn't sure since it looked like the last activity was October of last year.

**Comment by acao** [comment]
@mskelton I was about to get back go this!

**Pull request closed without merging**